### ISSUE TITLE
[New Concept Exercise] : functools

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2366]
This issue describes how to implement  the  `functools` concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach an understanding/use of `functools` (e.g,  the `functools module`)  in Python. 

## Learning objectives

Learn more about the functional tools the Python Standard Library provides through the `functools` module.
Build and understanding of and use the following methods and decorators from the module:

- `functools.partial()`
  - `partial.func`
  - `partial.args`
  - `partial.keywords`
- `functools.update_wrapper()`
-  `functools.cache()`
- ~~`@functools.cached_property()`~~ (_this seems better in `class_customization`_)
- `@functools.lru_cache()`
- ~~`@functools.total_ordering`~~ (_this method seems more appropriate in the `rich comparisons` exercise_)
- `@functools.singledispatch`
  - `@<function>.register()`
- `@functools.wraps()`
-  *class* `functools.partialmethod()` 
-  *class* `functools.singledispatchmethod()`

  

## Out of scope

- `classes`  & `class customization` beyond the direct use of the _class_ methods in this module.
- `comprehensions`
- `comprehensions` in `lambdas`
-  `decorators` (these have their own exercise.  See issue #2356 )
- `map()`, `filter()` or `functools.reduce()` in a `comprehension`
-  `functools.reduce()`(_this was already covered with `map()` and `filter()`_)
- `generators`
- using an `assignment expression` or "walrus" operator (`:=`) in a `lambda`
- class decorators beyond the ones described in this module.
- `enums`

  

## Concepts

- functional tools in python
- `functools` module
- `generic functions`
- `decorators`
- `higher-order functions`
- `partial objects` in python/`partial evaluation` in python
- `single dispatch`

  

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `bools`
- `classes`
- `class-customization`
- `class-components`
- `comparisons`
- `rich-comparisons`
- `decorators`
- `descriptors`
- `dicts`
- `dict-methods`
- `functions`
- `function-arguments`
- `higher-order-functions`
- `iteration`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

  

## Resources to refer to

- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Python Docs: functional programming HOWTO](https://docs.python.org/3/howto/functional.html)
- [Python Docs: functools module](https://docs.python.org/3/library/functools.html)
- [Pthon Module of the Week:  `functools` - Tools for Maniputlating Fuctions](https://pymotw.com/3/functools/)
- [Florian Dahlitz: Introduction to Python's Functools Module](https://florian-dahlitz.de/blog/introduction-to-functools)
- [PyDanny: Python Partials are Fun!](https://www.pydanny.com/python-partials-are-fun.html)
- [Composing Programs: Higher-Order Functions](https://composingprograms.com/pages/16-higher-order-functions.html)
- [Trey Hunner:  Ist it a Class or a Function?  It's a callable!](https://treyhunner.com/2019/04/is-it-a-class-or-a-function-its-a-callable/)
- [built-ins: Python Docs](https://docs.python.org/3/library/functions.html)
  - [`map()`](https://docs.python.org/3/library/functions.html#map) & [`filter()`](https://docs.python.org/3/library/functions.html#filter)
  - [`any()`](https://docs.python.org/3/library/functions.html#any) & [`all()`](https://docs.python.org/3/library/functions.html#all)
  - [`min()`](https://docs.python.org/3/library/functions.html#min) & [`max()`](https://docs.python.org/3/library/functions.html#max)
  - [`sum()`](https://docs.python.org/3/library/functions.html#sum) 
- [Real Python: Functional Programming in Python: When and How to Use it](https://realpython.com/python-functional-programming/)

  

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docshintsmd)

  * You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson) & [concept exercises](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md)

* ### Concept `about.md`

  **Concept file/issue**: The working copy of these files can be found here: [functools docs](https://github.com/exercism/python/tree/main/concepts/functools).  These may need editing to match the exercise writers needs/style.

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-introductionmd)

  - This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [Exercise `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-docsintroductionmd)

  - This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** :  ``
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#metadata-files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
- `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

- Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.
- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by Tarun193** [comment]
Hey, I just want to work onthis . But I am a newbee to open source. So, may I try to built it?

**Comment by BethanyG** [comment]
Hi @Tarun193 👋🏽 

We're delighted you'd like to take this on!  🎉  

I'll assign this issue to you and remove the [help wanted] tag. 

And don't worry -- we're happy to help you through the process.  You can always ask questions here (_or on Slack or Gitter if you are on those platforms_), and we have loads of documents that outline what's required.  Make sure to go over those spec docs carefully.  😅   But at the same time, don't get intimidated - it if you take the tasks one at a time and work at a steady pace, the whole thing will come together nicely. 

Additionally, I am happy to write up the concept docs (_if you decide you don't want to do that part_) and provide feedback or collaborate with you prior to you submitting a PR for formal review.

Just ping with questions - and have fun exploring and writing!  😄  💛 💙

**Comment by Tarun193** [comment]
Thanks , I will let you know when I need some help.

**Comment by Tarun193** [comment]
Hey, I want to know that I have to write about these concepts.

functional tools in python
functools module
generic functions
decorators
higher-order functions
partial objects in python/partial evaluation in python
single dispatch

**Comment by Tarun193** [comment]
hey anyone is there

**Comment by kotp** [comment]
Yes.  I saw this "Hey, " message when it got posted.  It was, to me, a declaration, and so no request of action was recognized.

So, what can we help you with @Tarun193 ?

**Comment by BethanyG** [comment]
@Tarun193  -- My sincere apologies!  I have been really swamped with work, and this slipped through the cracks.  **Yes**, we do need to write about those concepts.  But if you'd like, I can do the concept write-up, and you can focus on the exercise itself.  Just let me know how you'd like to proceed.

Here is [an example](https://github.com/exercism/python/tree/main/concepts/classes) of what a concept write-up looks like.  There is a longer `about` document, and a shorter "summary" with the `introduction`.  There is also a `links` document for learning links for students to explore for a deeper understanding.

By contrast, [here is an example](https://github.com/exercism/python/tree/main/exercises/concept/ellens-alien-game) of the related concept exercise.  Note that there is an introduction -- but it can be pulled from the concept docs if you don't want to do a custom one.

We try to focus the concept docs on the topics that the exercise also covers.  The `introduction` is available before a student completes the concept exercise, and the `about` displays after.  The about often also covers additional details and applications of the concepts, and maybe some thought-provoking points or topics.

Beyond that, it's helpful to go through the links at the top of this issue for specifications, requirements, and other how-tos.  I believe they're fairly up to date.

Let me know if you have additional questions or issues.  I won't have time before this weekend, but can work on a concept writeup draft then, if you don't want to write one.

Best,

**Comment by Tarun193** [comment]
No problem @BethanyG  ,I would like to help in writing concept. I will try  to do as good as I can.

**Comment by Tarun193** [comment]
hey I have wrote the introduction part of the concept , please let me know if some improvements required.

**Comment by kotp** [comment]
#2986

**Comment by BethanyG** [comment]
❗❗  **`Note to the contributor who picks this task:`**

The concept docs for this exercise were completed by a previous contributor, and can be found here:  [functools concept docs](https://github.com/exercism/python/tree/main/concepts/functools).  

We suggest going through them before getting started with the rest of the documents and pinging if you have any questions or issues.

**Comment by Metallifax** [comment]
Hello :wave: 
I can pick up from where the last author left off since I can see there is already some momentum going. I'll study the resources while I wait to hear back. Thank you for your time and regards.

**Comment by BethanyG** [comment]
All yours and many thanks @Metallifax!  🌟 🎉 🌮

**Comment by BethanyG** [comment]
@Metallifax  - do you still want to work on this?  Totally OK if you do not.  Just ping when you have time, and let us know.  Thanks!

**Comment by Metallifax** [comment]
@BethanyG I feel bad for leaving this claimed for so long, but I've also been focused on other projects/work/study. That said, I don't want to be all excuses and would like to make a solid commitment to moving this exercise forward.

I finish schoolwork/tests on June 5th, and I'm hoping to have my personal project finished by then (hopefully); would that be an okay timeline to start making progress on this? 

If this is too far into the future, I don't mind bowing out and letting another contributor take over. As easy as it'd be to just offload the work, I think the experience of making a concept exercise would be a tremendous opportunity for growth and would love to see the concept make it to production.

Let me know if this timeline works for you guys. Regards.

**Comment by BethanyG** [comment]
@Metallifax  -- Thanks for letting us know!  Totally understandable that you have life, studying, and work!  I don't think we're in any particular hurry, so I'm fine with leaving you assigned (_I just wanted to make sure you still wanted to take this on_).  If someone else does pick this up, we can ping at that point, or set up a collaboration.  😄 

_Good luck with school and all your projects!_

**Comment by Metallifax** [comment]
I am working on it, just slower than I'd like. I'll try to get a PR out by the weekend to move this along. Thanks for the patience on this.

**Comment by BethanyG** [comment]
@Metallifax - This issue is up for a re-write, so I might close this and then re-open it fresh, if that's OK with you?  I'll ping you on Slack whenI am done and you can then comment so I can re-assign you.

**Comment by Metallifax** [comment]
Go ahead! Since we're on the topic of adding resources, I found broken links to the concept and concept exercise docs in the issue description (you're probably already aware):

- https://github.com/exercism/docs/blob/main/anatomy/tracks/concept-exercises.md
- https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md

Which I think should go to:

- https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md
- https://github.com/exercism/docs/blob/main/building/tracks/concepts.md

Unless we're going straight to the frontend:

- https://exercism.org/docs/building/tracks/concept-exercises
- https://exercism.org/docs/building/tracks/concepts

Thanks again.

**Comment by BethanyG** [comment]
Yup!  😄   That's one of the reasons we're closing most of the issues logged in Feb. and re-writing them!  See #3091  and #3092 for some examples.  In addition to updating the links, we're going to be splitting the concept docs writing from the _exercise_ writing, so if you still want to do both things, remember comment on both issues and I'll assign you to them.

**Comment by Metallifax** [comment]
Thank you for summarizing, and sounds great! I'll keep plugging away and keep an eye out for the new issue in the meantime.

**Comment by BethanyG** [comment]
The new issue is HERE:  https://github.com/exercism/python/issues/3097.  😄   There will be an additional improvement issue for the concept docs shortly.  Closing this now as "OLD".
